src/Nethermind/Nethermind.TxPool/Filters/BalanceZeroFilter.cs

```diff
@@ -42,6 +48,24 @@ public AcceptTxResult Accept(Transaction tx, TxFilteringState state, TxHandlingO
                     AcceptTxResult.InsufficientFunds.WithMessage($"Balance is {balance} less than sending value {tx.Value}");
             }
 
+            IReleaseSpec spec = _specProvider.GetCurrentHeadSpec();
+            UInt256 affordableGasPrice = tx.CalculateGasPrice(spec.IsEip1559Enabled, _headInfo.CurrentBaseFee);
+            if (UInt256.AddOverflow(affordableGasPrice, tx.Value, out UInt256 txCostAndValue))
```

Comment:
Cost of transaction is `value + GasPrice(FeePerGas) * GasLimit`.
Here you are checking `value + GasPrice`, which is not multiplying `FeePerGas * GasLimit` which is not correct.

Second thing, we should use `MaxFeePerGas` here - tx must be able to afford `value + MaxFeePerGas * GasLimit`. Using `CalculateGasPrice` (so effective gas price) can result in adding tx which is latent invalid - we are doing this check anyway in `BalanceTooLowFilter`, but if you decided to add fast check, we should use the same logic here, otherwise fast check will simple accept txs with `balance >= value + effectiveGasPrice * GasLimit` but with `balance < value + MaxFeePerGas * GasLimit` and such txs will be rejected later in `BalanceTooLowFilter`. Also we can drop `headInfo` as base fee will not matter anymore.

Checking effective gas price is fine in filters checking fee, but in the ones checking balance we should require account to be able to pay max declared value

Reply:
Reverted last commit :)